fabric-events-interaction-v0/src/main/java/net/fabricmc/fabric/api/event/player/PlayerBlockBreakEvents.java

```diff
@@ -0,0 +1,123 @@
+import net.minecraft.util.math.BlockPos;
+import net.minecraft.world.World;
+
+import net.fabricmc.fabric.api.event.Event;
+import net.fabricmc.fabric.api.event.EventFactory;
+
+public final class PlayerBlockBreakEvents {
+	private PlayerBlockBreakEvents() { }
+
+	/**
+	 * Callback before a block is broken.
+	 * Only called on the server, however updates are synced with the client.
+	 *
+	 * <p>If any listener cancels a block breaking action, that block breaking
+	 * action is cancelled and {@link CANCELED} event is fired. Otherwise, the
+	 * {@link AFTER} event is fired.</p>
+	 */
+	public static final Event<Before> BEFORE = EventFactory.createArrayBacked(Before.class,
+			(listeners) -> (world, player, pos, state, entity) -> {
+				for (Before event : listeners) {
+					boolean result = event.beforeBlockBreak(world, player, pos, state, entity);
+
+					if (!result) {
+						return false;
+					}
+				}
+
+				return true;
+			}
+	);
+
+	/**
+	 * Callback after a block is broken.
+	 *
+	 * <p>Called on the Server only.</p>
+	 */
+	public static final Event<After> AFTER = EventFactory.createArrayBacked(After.class,
+			(listeners) -> (world, player, pos, state, entity) -> {
+				for (After event : listeners) {
+					event.afterBlockBreak(world, player, pos, state, entity);
+				}
+			}
+	);
+
+	/**
+	 * Callback when a block break has been canceled.
+	 *
+	 * <p>Called on the logical server only. May be used to send packets to revert client-side block changes.</p>
+	 */
+	public static final Event<Canceled> CANCELED = EventFactory.createArrayBacked(Canceled.class,
+			(listeners) -> (world, player, pos, state, entity) -> {
+				for (Canceled event : listeners) {
+					event.onBlockBreakCancel(world, player, pos, state, entity);
+				}
+			}
+	);
+
+	@FunctionalInterface
+	public interface Before {
+		/**
+		 * Called before a block is broken and allows cancelling the block breaking.
+		 *
+		 * <p>Implementations should not modify the world or assume the block break has completed or failed.</p>
+		 *
+		 * @param world the world in which the block is broken
+		 * @param player the player breaking the block
+		 * @param pos the position at which the block is broken
+		 * @param state the block state <strong>before</strong> the block is broken
+		 * @param entity the block entity <strong>before</strong> the block is broken, can be {@code null}
+		 * @return {@code false} to cancel block breaking action, or {@code true} to pass to next listener
+		 */
+		boolean beforeBlockBreak(World world, PlayerEntity player, BlockPos pos, BlockState state, /* Nullable */ BlockEntity entity);
+	}
+
+	@FunctionalInterface
+	public interface After {
+		/**
+		 * Called after a block is successfully broken.
+		 *
+		 * @param world the world where the block was broken
+		 * @param player the player who broke the block
+		 * @param pos the position where the block was broken
+		 * @param state the block state <strong>after</strong> the block was broken
```

Comment:
Wouldn't it be more useful to provide the state of the block that was broken as anyone who wants to query the currently state can get it from the world.

Or is this a mistake in javadoc as the the state from impl seems to be from local capture

Reply:
The state is from before the block is broken, thanks for saying, changing javadoc now